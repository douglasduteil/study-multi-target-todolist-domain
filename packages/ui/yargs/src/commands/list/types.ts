//

import { Arguments } from "yargs";

//

export type ListArguments = Arguments<{ [key in keyof ListOptions]: boolean }>;
export type Aliases = "ls" | "la" | "ll";

export interface ListOptions {
  all: boolean;
  // XXX(douglasduteil): Move json out of here
  // json should be part of the "list" filter as it's format
  json: boolean;
  long: boolean;
}
