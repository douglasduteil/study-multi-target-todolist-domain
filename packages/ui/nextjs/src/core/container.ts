//

import { Container } from "inversify";
import { createContext } from "react";

//

export const container = new Container();
export const ContainerContext = createContext<Container>(container);
