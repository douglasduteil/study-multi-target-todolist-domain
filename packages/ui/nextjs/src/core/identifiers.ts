//

const prefix = "@todolist/packages/ui/nextjs";

export const IDENTIFIER = {
  CORE_FILTER: Symbol.for(`${prefix}#CORE_FILTER`),
  DATA_SOURCE: Symbol.for(`${prefix}#DATA_SOURCE`)
};
