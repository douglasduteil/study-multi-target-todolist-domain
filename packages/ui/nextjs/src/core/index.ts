//

export * from "./container";
export * from "./identifiers";
