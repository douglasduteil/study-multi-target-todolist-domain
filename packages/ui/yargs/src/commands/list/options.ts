//

import { Options } from "yargs";
import { Aliases, ListOptions } from "./types";

//

export const group = "List Options:";
export const ALIASES_DEFAULT_OPTIONS: {
  [key in Aliases]: Partial<ListOptions>
} = {
  la: { all: true },
  ll: { long: true },
  ls: {}
};

export const options: { [key in keyof ListOptions]: Options } = {
  all: {
    alias: "a",
    default: false,
    describe: "Show done todos that are normally hidden",
    type: "boolean"
  },
  json: {
    default: false,
    describe: "Show information as a JSON array",
    type: "boolean"
  },
  long: {
    alias: "l",
    default: false,
    describe: "Show extended information",
    type: "boolean"
  }
};
