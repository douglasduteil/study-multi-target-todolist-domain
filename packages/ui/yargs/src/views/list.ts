//

import { Todo } from "@todolist/core";
import dedent from "dedent";
import ago from "s-ago";
import { TodoWithoutMeta } from "../interfaces";

//

export function listView(todos: ReadonlyArray<Todo | TodoWithoutMeta>): string {
  return todos
    .map(template)
    .reduce((view, line) => `${view}\n${line}`, "")
    .trim();
}

//

function isTodo(todo: Todo | TodoWithoutMeta): todo is Todo {
  return Boolean((todo as Todo).createdAt);
}

function template(todo: Todo | TodoWithoutMeta): string {
  return dedent`
    ${idSymbol(todo.id)} ${completeSymbol(todo.completed)} ${todo.title} ${
    isTodo(todo) ? metaInfo(todo) : ""
  }
  `.trim();
}

function idSymbol(id?: string): string {
  return `(${id})`;
}

function completeSymbol(complete?: boolean): string {
  return complete ? "☒" : "☐";
}

function metaInfo(todo: Todo): string {
  return (
    "\n  " +
    dedent`
      Created ${ago(todo.createdAt)}
    `.trim()
  );
}
