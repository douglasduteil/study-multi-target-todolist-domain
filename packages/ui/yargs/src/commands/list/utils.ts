//

import { ALIASES_DEFAULT_OPTIONS } from "./options";
import { Aliases, ListArguments } from "./types";

//

export function mapAliasOnOptions(argv: ListArguments): ListArguments {
  const alias = argv._[0] as Aliases;
  const options = ALIASES_DEFAULT_OPTIONS[alias];
  return { ...argv, ...options };
}
