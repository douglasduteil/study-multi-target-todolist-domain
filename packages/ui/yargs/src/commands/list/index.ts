//

import { TodoDataSource, TodoListFilterFn } from "@todolist/core";
import pipe from "ramda/es/pipe";
import { Argv, CommandModule } from "yargs";
import { jsonView, listView } from "../../views";
import { group, options } from "./options";
import { Aliases, ListArguments } from "./types";
import { mapAliasOnOptions } from "./utils";

//

export class ListCommand implements CommandModule<Argv<{}>, ListArguments> {
  public command = "list [options]";
  public describe = "List todos";
  public aliases: ReadonlyArray<Aliases> = ["ls", "la", "ll"];
  constructor(
    private todolistFilterFn: TodoListFilterFn,
    private logger: Console,
    private todoDataSource: TodoDataSource
  ) {}

  // HACK(douglasduteil): CommandBuilder<any, any> ...
  // I didn't find how to make tsc happy without any-s here...
  public builder(yargs: Argv<{}>): Argv<any> {
    return yargs.options(options).group(Object.keys(options), group);
  }

  // ? NOTE(douglasduteil): `handler` is assigned here to preserve the `this`
  // ? `yargs` seems to call the handlers with forced `.bind(undefined, [...])`
  public handler = async (args: ListArguments) => {
    const opts = mapAliasOnOptions(args);

    const transform = this.todolistFilterFn(opts);

    const render = pipe(
      args.json ? jsonView : listView,
      this.logger.log
    );

    pipe(
      transform,
      render
    )(await this.todoDataSource.getTodos());
  };
}
