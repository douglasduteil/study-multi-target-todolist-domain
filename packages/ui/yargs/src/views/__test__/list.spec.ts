//

import { Todo } from "@todolist/core";
import { listView } from "../list";

//

test("should return list of 0 todos", () => {
  expect(listView([])).toMatchSnapshot();
});

test("should return list of one foo todo", () => {
  expect(
    listView([createFakeTodo({ id: "42", title: "foo" })])
  ).toMatchSnapshot();
});

//

function createFakeTodo(template: Partial<Todo>): Todo {
  return {
    completed: false,
    createdAt: new Date(0),
    id: "",
    title: "",
    updatedAt: new Date(0),
    ...template
  };
}
