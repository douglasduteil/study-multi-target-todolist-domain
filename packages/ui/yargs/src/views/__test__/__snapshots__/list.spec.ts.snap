// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return list of 0 todos 1`] = `""`;

exports[`should return list of one foo todo 1`] = `
"(42) ‚òê foo 
Created 49 years ago"
`;
